osquery/worker/ipc/posix/pipe_channel.cpp

```diff
@@ -113,22 +113,19 @@ Status PipeChannel::recvStringMessageImpl(std::string& message) {
   }
 
   message.resize(static_cast<size_t>(message_size));
-  result = read(read_pipe_fd,
-                reinterpret_cast<char*>(&message[0]),
-                static_cast<size_t>(message_size));
-
-  if (result < 0) {
-    return Status::failure(
-        errno,
-        "Failed to read the size of the message from the pipe of table " +
-            table_name_ + ", errno " + std::to_string(errno));
-  }
-
-  if (result != message_size) {
-    return Status::failure(
-        "Failed to read the entire message from the pipe of table " +
-        table_name_ + ", read only " + std::to_string(result) + "/" +
-        std::to_string(message_size) + " bytes");
+  ssize_t pos = 0;
+  while (pos < message_size) {
+    result = read(read_pipe_fd,
+                  reinterpret_cast<char*>(&message[pos]),
+                  static_cast<size_t>(message_size - pos));
+
+    if (result < 0) {
+      return Status::failure(
+          errno,
+          "Failed to read the message from the pipe of table " + table_name_ +
+              ", errno " + std::to_string(errno));
+    }
```

Comment:
You should also handle the case where the `result` is 0 because the pipe has been closed, otherwise this ends in an infinite loop.

Use this:
https://github.com/osquery/osquery/pull/7139/files#diff-83aebbdb912ef61c487ba416a097a16b2211d12006b3b21368f7e12f80da1e24R97-R100

Reply:
added.  Nice catch.